From f635b8c019b26ac0cf1b94a330c7ea401aaca111 Mon Sep 17 00:00:00 2001
From: Sahil Takiar <takiar.sahil@gmail.com>
Date: Wed, 6 Jul 2016 16:34:19 -0500
Subject: [PATCH 0664/1164] HIVE-14090: JDOExceptions thrown by the Metastore
 have their full stack trace returned to clients
 (Sahil Takiar, reviewed by Sergio Pena)

(cherry picked from commit 13e7029163287bcd55958e99ac2fd408863d6443)

Conflicts:
	ql/src/test/results/clientnegative/create_with_constraints_duplicate_name.q.out

Change-Id: I5242436f65c2cc2c9e48f7ebca518f1930e5f4c1
---
 .../hadoop/hive/metastore/RetryingHMSHandler.java  |    6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/metastore/src/java/org/apache/hadoop/hive/metastore/RetryingHMSHandler.java b/metastore/src/java/org/apache/hadoop/hive/metastore/RetryingHMSHandler.java
index 76c698c..334f23d 100644
--- a/metastore/src/java/org/apache/hadoop/hive/metastore/RetryingHMSHandler.java
+++ b/metastore/src/java/org/apache/hadoop/hive/metastore/RetryingHMSHandler.java
@@ -193,9 +193,9 @@ public Result invokeInternal(final Object proxy, final Method method, final Obje
 
       if (retryCount >= retryLimit) {
         LOG.error("HMSHandler Fatal error: " + ExceptionUtils.getStackTrace(caughtException));
-        // Since returning exceptions with a nested "cause" can be a problem in
-        // Thrift, we are stuffing the stack trace into the message itself.
-        throw new MetaException(ExceptionUtils.getStackTrace(caughtException));
+        MetaException me = new MetaException(caughtException.getMessage());
+        me.initCause(caughtException);
+        throw me;
       }
 
       assert (retryInterval >= 0);
-- 
1.7.9.5

